### PR TITLE
Add Demo for EC keygen

### LINKED ISSUE #14112: Add an EVP demo for EC key generation
Use a named curve and include a validation step as a separate function.

It should be the current best practice (fetch), have proper error handling and resource allocation/deallocation and be in the style of a known answer test.

### PR DESCRIPTION
**Pull request by slontis**
Fixes #14112

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: demos/pkey/EVP_PKEY_EC_keygen.c [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+static int get_key_values(EVP_PKEY *pkey);
+
+/*
+ * The following code shows how to generate a EC key from a curve name
+ * with additional parameters. If only the curve name is required then the
+ * simple helper can be used instead i.e. Either
+ * pkey = EVP_EC_gen(curvename); OR
+ * pkey = EVP_PKEY_Q_keygen(libctx, propq, "EC", curvename);
+ */
+static EVP_PKEY *do_ec_keygen(void)
```
Comment (paulidale) [inline demos/pkey/EVP_PKEY_EC_keygen.c:30]: I'm unsure if the function being static makes sense or not.

File: demos/pkey/EVP_PKEY_EC_keygen.c [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        goto cleanup;
+    }
+
+    if (!EVP_PKEY_get_octet_string_param(pkey, OSSL_PKEY_PARAM_PUB_KEY,
+                                        out_pubkey, sizeof(out_pubkey), &len)) {
+        fprintf(stderr, "Failed to get public key\n");
+        goto cleanup;
+    }
+
+    out_priv = BN_new();
```
Comment (paulidale) [inline demos/pkey/EVP_PKEY_EC_keygen.c:109]: Doesn't `EVP_PKEY_get_bn_param()` do the allocation for us?  I.e. this needn't be done separately?

**Review by paulidale — COMMENTED** [review]
Looks good.  Some stylistic comments.  We should create some guidelines for demos.  What is best practice, what can/should be avoided etc.

File: demos/pkey/EVP_PKEY_EC_keygen.c (hunk 2 above)
Comment (slontis) [reply demos/pkey/EVP_PKEY_EC_keygen.c:109]: I dont think so..

Comment (paulidale) [reply demos/pkey/EVP_PKEY_EC_keygen.c:109]: If it calls down to the OSSL_PARAM_get_BN function it will (assuming the pointer is initialised to NULL).

**Comment by paulidale** [comment]
Probably should have a Fixes #14112 note

File: demos/pkey/EVP_PKEY_EC_keygen.c (hunk 1 above)
Comment (slontis) [reply demos/pkey/EVP_PKEY_EC_keygen.c:30]: Its a local function.. Why would it not be static?

**Comment by slontis** [comment]
Last force push was to fix the commit message.

File: demos/pkey/EVP_PKEY_EC_keygen.c (hunk 1 above)
Comment (paulidale) [reply demos/pkey/EVP_PKEY_EC_keygen.c:30]: It's a demo and this is the function that will be used (i.e. exported)

File: demos/pkey/Makefile [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#
+# To run the demos when linked with a shared library (default):
+#
+#    LD_LIBRARY_PATH=../.. ./EVP_PKEY_EC_keygen
```
Comment (paulidale) [inline demos/pkey/Makefile:4]: Would `../../util/wrap.pl EVP_PKEY_EC_keygen` make more sense?  This is doing more than just setting the LD_LIBRARY_PATH.

**Review by paulidale — APPROVED** [review]
LGTM.  One suggestion about the running instruction.

File: demos/pkey/Makefile (hunk 3 above)
Comment (slontis) [reply demos/pkey/Makefile:4]: Maybe.. Probably better to do that as a seperate pass over all the demos

**Comment by slontis** [comment]
Thanks. Merged to master.

**Pull request closed without merging**